### ISSUE TITLE
Just notifying about warnings when in CI

### ISSUE THREAD

**Issue by underyx** [coala/coala#2897]
There's a bunch of bears that are only around 95% reliable, or maybe we just don't know how to configure it to make it just strict enough to not bother us about too many things. The problem is that the only way to deal with this is to just disable them — if they fail the CI build for some bullshit reason, that'll be annoying, if instead we somehow run the checks and mute them (make them not fail the build), no one will ever see the report of warnings, since no one ever checks succeeded build logs.

So, how about we specify an alternative communication channel for warnings?
## Mockup

A mockup with very little effort put into it would go like:
1. Set `COALA_SLACK_WEBHOOK_URL` as a CI env variable with your secret webhook URL
2. Configure a bear:
   
   ``` ini
   [english]
   bears = WriteGoodBear
   default_action = WarnAction  # we don't need a literary masterwork
   
   [scary-keywords]
   bears = KeywordBear
   ci_keywords = DROP TABLE
   default_action = WarnAction  # we don't need a literary masterwork
   ```
3. :tada:! You will now receive a slack message whenever WriteGoodBear gives you a result on CI.
## Moonshots
- While this would already work, all the reports on the results would get old quick, how do we ignore seen warnings? Without that it'd be almost useless
- Could we check `git blame` and tag the person who committed the warned code the slack message?
- More connectors, such as SMTP, gitter, GitHub/GitLab line comments

**Comment by Nosferatul** [comment]
I'm positive that we need this. Sometimes failed builds can be quite annoying

**Comment by plaindocs** [comment]
| While this would already work, all the reports on the results would get old quick, how do we ignore seen warnings? Without that it'd be almost useless

would be the issue that I'd see. Otherwise it circles back to having to see it / ignoring it. If your slack gets swamped with same error messages constantly people are just going to turn them off or ignore them.

**Comment by underyx** [comment]
@plaindocs Definitely. I believe someone (@sims1253 maybe?) already has experience with fancy code chunk matching magic, so we _should_ be able to implement it reasonably well. I was evil enough to categorize this as a moonshot so that the issue seems less daunting and is more likely to be picked up, but it really would be nearly useless without this.

By the way, awesome to see you snooping around the coala project! You probably don't remember, but we met briefly in Budapest twice after Write the Docs EU 2014 👯‍♂️

**Comment by plaindocs** [comment]
That would definitely be something I'd like to see!

I have a couple more ideas up my sleeve for when I meet Lasse in Berlin this coming weekend. :-D

Sure I remember you from Budapest! I'm using Coala on the Travis CI docs, and planning to reuse some of that on the WTD site as well.

**Comment by sims1253** [comment]
@underyx You want me to magic the git blame for code that has a warning attached to it?

I haven't really worked  with git blame a lot but as we have a file and line number from the warnings/results it should be rather easy as long as git offers an author.

My thought would be to either fail a build because it does not satisfy my demands, including warnings and stuff, or let it pass and then live with what I ignored. But I can see why you would want a build pass with warnings if you want faster progress and less quality. Maybe just posting the warnings to some api would be the easiest? Why is a blame needed?

**Comment by underyx** [comment]
@sims1253

>But I can see why you would want a build pass with warnings if you want faster progress and less quality.

You might be misunderstanding my case here. See this:

```
[scary-keywords]
bears = KeywordBear
ci_keywords = DROP TABLE
default_action = WarnAction  # it might be legit needed
```

You might describe this feature more like something that draws attention to things requiring careful review. If that `DROP TABLE` query passed review, I don't want to warn about it again.

**Comment by sims1253** [comment]
@underyx ok I totally misread what you wanted.
Any idea how to store the information if a warning was issued already? Maybe you could create an issue and check if it exists?

**Comment by underyx** [comment]
No plans yet at all. Using an issue tracker sounds like it'd quickly get out of hand as people would request 50 different integrations for their trackers. How about writing to a file and leaving it up the user to set up CI to persist that across builds?

**Comment by plaindocs** [comment]
Caching is not quite the right fit though, as the file potentially changes any time you edit it, so you need to retrieve it per run (and probably per user?) if there is match.

**Comment by plaindocs** [comment]
Looks like Alex looks at diffs only if run on Travis CI.  https://github.com/wooorm/alex#workflow

**Comment by wooorm** [comment]
alex can do that (through the `--diff` flag), but not by default!
